File: QuasselDroidNG/src/main/java/de/kuschku/libquassel/primitives/serializers/StringSerializer.java
package de.kuschku.libquassel.primitives.serializers;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;
import java.nio.charset.Charset;

public class StringSerializer implements PrimitiveSerializer<String> {
    private static final StringSerializer serializer = new StringSerializer();

    private StringSerializer() {
    }

    public static StringSerializer get() {
        return serializer;
    }

    @Override
    public void serialize(@NonNull final ByteChannel channel, @Nullable final String data) throws IOException {
        if (data == null) {
            IntSerializer.get().serialize(channel, 0xffffffff);
        } else {
            final ByteBuffer contentBuffer = Charset.forName("UTF-16BE").encode(data);
            IntSerializer.get().serialize(channel, contentBuffer.limit());
            channel.write(contentBuffer);
        }
    }

    @Nullable
    @Override
    public String deserialize(final ByteBuffer buffer) throws IOException {
        final int len = IntSerializer.get().deserialize(buffer);
        if (len == 0xffffffff)
            return null;
        else {
            final ByteBuffer contentBuffer = ByteBuffer.allocate(len);
            contentBuffer.put(buffer.array(), buffer.position(), len);
            contentBuffer.position(0);
            buffer.position(buffer.position() + len);
            return Charset.forName("UTF-16BE").decode(contentBuffer).toString();
        }
    }
}

File: app/src/main/java/de/kuschku/quasseldroid_ng/protocol/primitive/serializer/StringSerializer.kt
package de.kuschku.quasseldroid_ng.protocol.primitive.serializer

import de.kuschku.quasseldroid_ng.protocol.QVariant
import de.kuschku.quasseldroid_ng.protocol.Quassel_Features
import de.kuschku.quasseldroid_ng.util.helpers.copyTo
import de.kuschku.quasseldroid_ng.util.nio.ChainedByteBuffer
import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.Charset
import java.nio.charset.CharsetDecoder
import java.nio.charset.CharsetEncoder

abstract class StringSerializer(
  private val encoder: CharsetEncoder,
  private val decoder: CharsetDecoder,
  private val trailingNullBytes: Int
) : Serializer<String?> {
  constructor(charset: Charset,
              trailingNullByte: Boolean = false) : this(
    charset.newEncoder(),
    charset.newDecoder(),
    if (trailingNullByte) {
      1
    } else {
      0
    }
  )

  object UTF16 : StringSerializer(Charsets.UTF_16BE)
  object UTF8 : StringSerializer(Charsets.UTF_8)
  object C : StringSerializer(Charsets.ISO_8859_1, trailingNullByte = true)

  override fun serialize(buffer: ChainedByteBuffer, data: String?, features: Quassel_Features) {
    if (data == null) {
      IntSerializer.serialize(buffer, -1, features)
    } else {
      val charBuffer = CharBuffer.allocate(data.length)
      charBuffer.put(data)
      charBuffer.flip()
      val byteBuffer = encoder.encode(charBuffer)
      IntSerializer.serialize(buffer, byteBuffer.remaining() + trailingNullBytes, features)
      buffer.put(byteBuffer)
      for (i in 0 until trailingNullBytes)
        buffer.put(0)
    }
  }

  fun serialize(data: String?): ByteBuffer = if (data == null) {
    ByteBuffer.allocate(0)
  } else {
    val charBuffer = CharBuffer.allocate(data.length)
    charBuffer.put(data)
    charBuffer.flip()
    encoder.encode(charBuffer)
  }

  fun deserializeAll(buffer: ByteBuffer): String? {
    val len = buffer.remaining()
    return if (len == -1) {
      null
    } else {
      val byteBuffer = ByteBuffer.allocate(len)
      buffer.copyTo(byteBuffer)
      byteBuffer.clear()
      byteBuffer.limit(byteBuffer.limit() - trailingNullBytes)
      decoder.decode(byteBuffer).toString()
    }
  }

  override fun deserialize(buffer: ByteBuffer, features: Quassel_Features): String? {
    val len = IntSerializer.deserialize(buffer, features)
    return if (len == -1) {
      null
    } else {
      val byteBuffer = ByteBuffer.allocate(len)
      buffer.copyTo(byteBuffer)
      byteBuffer.clear()
      byteBuffer.limit(byteBuffer.limit() - trailingNullBytes)
      decoder.decode(byteBuffer).toString()
    }
  }
}

fun QVariant<ByteBuffer>?.deserializeString(
  serializer: StringSerializer) = if (this?.data == null) {
  null
} else {
  serializer.deserializeAll(data)
}

fun ByteBuffer?.deserializeString(serializer: StringSerializer) = if (this == null) {
  null
} else {
  serializer.deserializeAll(this)
}

fun String?.serializeString(serializer: StringSerializer) = if (this == null) {
  null
} else {
  serializer.serialize(this)
}
